deps/chakrashim/src/jsrtcontextshim.cc

```diff
@@ -368,11 +368,34 @@ bool ContextShim::InitializeProxyOfGlobal() {
   });
 }
 
-bool ContextShim::EnsureInitialized() {
+
+bool ContextShim::InitializeCurrentContextShim() {
   if (initialized) {
     return true;
   }
 
+  JsContextRef oldContext;
+  if (JsGetCurrentContext(&oldContext) != JsNoError) {
+    return false;
+  }
+  JsContextRef thisContext = GetContextRef();
+  if (JsSetCurrentContext(thisContext) != JsNoError) {
+    return false;
+  }
+  if (!EnsureInitialized()) {
+    Fatal("Failed to initialize context");
+  }
+  if (JsSetCurrentContext(oldContext) != JsNoError) {
+    return false;
+  }
+  return true;
+}
+
+bool ContextShim::EnsureInitialized() {
+  if (initialized) {
+    return true;
+  }
+  initializing = true;
```

Comment:
I kind of see why you adding "initializing". But it looks it is almost effectively equivalent to moving "initialized = true" here. Because once you set "initializing = true" here, all the getters afterwards will directly get the fields. It doesn't provide much value. I think you can refactor like suggested (EnsureInitialize call DoInitialize). Move "initialized = true" here, then call DoInitialize, if that returns false (fail for any reason), Fatal error... or revert initialized = false and return.

Then all your getter checks only need to check one "initialized" and call EnsureIntialized if not. You can even do more refactoring, so that all the getters can even just call EnsureInitialized which checks "initialized", and make EnsureInitialized a small function and inline.